src/main/java/com/esri/samples/localserver/local_server_services/LocalServerServicesController.java

```diff
@@ -0,0 +1,228 @@
+            localService = new LocalFeatureService(serviceUrl);
+            break;
+          case "Geoprocessing Service":
+            serviceUrl = Paths.get(getClass().getResource("/local_server/MessageInABottle.gpk").toURI()).toString();
+            localService = new LocalGeoprocessingService(serviceUrl);
+        }
+
+        startLocalService(localService, serviceSelected);
+
+      } else {
+        // if service is currently running on server warn user
+        Platform.runLater(() -> {
+          Alert dialog = new Alert(AlertType.INFORMATION);
+          dialog.setHeaderText("Service Running");
+          dialog.setContentText(serviceSelected + " has already been started.");
+          dialog.showAndWait();
+        });
+      }
+    }
+  }
+
+  /** 
+   * Starts a Local Service.
+   * <p>
+   * The status of the Local Service is displayed in the center text area of the application.
+   * The URL of the Local Service is added to the list of running services.
+   * 
+   * @param localService Local Service to start.
+   * @param serviceName Name of Local Service. 
+   */
+  private void startLocalService(LocalService localService, String serviceName) {
+    localService.addStatusChangedListener(status -> {
+      statusLog.appendText(serviceName + " Status: " + status.getNewStatus().toString() + "\n");
+      if (status.getNewStatus() == LocalServerStatus.STARTED) {
+        Platform.runLater(() -> runningServices.getItems().add(serviceName + " URL ->  " + localService.getUrl()));
+      }
+    });
+    localService.startAsync();
+  }
+
+  /**
+   * Gets the Local Service that is selected from the list of running services and stops it.
+   * 
+   * @param event the event that fired this action
+   */
+  @FXML
+  protected void handleStopLocalService(ActionEvent event) {
+
+    int serviceIndex = runningServices.getSelectionModel().getSelectedIndex();
+    services.get(serviceIndex).stopAsync();
+    runningServices.getItems().remove(serviceIndex);
+  }
+
+  /** 
+   * Enables the URL and Stop Service button when a service is selected. 
+   */
+  @FXML
+  protected void enableServiceButtons() {
+
+    int listIndex = runningServices.getSelectionModel().getSelectedIndex();
+    if (listIndex != -1) {
+      btnStopService.setDisable(false);
+      btnURL.setDisable(false);
+    } else {
+      btnStopService.setDisable(true);
+      btnURL.setDisable(true);
+    }
+  }
+
+  /** 
+   * Opens a browser to the Local Service that is selected in the list of running services.
+   * 
+   * @param event event that fired this action 
+   */
+  @FXML
+  protected void handleURL(ActionEvent event) {
+    String url = runningServices.getSelectionModel().getSelectedItem().split(">")[1].trim();
+    hostServices.showDocument(url);
+  }
+
+  /** 
+   * Stops any Local Services that are currently running and then stops the Local Server. 
+   */
+  void terminate() {
+    if (server != null && server.getStatus() == LocalServerStatus.STARTED) {

```

Comment:
Move this logic to `handleStopLocalServer` and call it from `terminate` since this logic is actually doing the stopping.